api/envoy/config/route/v3/route_components.proto

```diff
@@ -904,13 +904,14 @@ message RouteAction {
     // for additional documentation.
     WeightedCluster weighted_clusters = 3;
 
-    // [#not-implemented-hide:]
-    // Name of the cluster specifier plugin to use to determine the cluster for
-    // requests on this route. The plugin name must be defined in the associated
-    // :ref:`envoy_v3_api_field_config.route.v3.RouteConfiguration.cluster_specifier_plugins`
-    // in the
-    // :ref:`envoy_v3_api_field_config.core.v3.TypedExtensionConfig.name` field.
-    string cluster_specifier_plugin = 37;
+    // Name of the cluster provider to use to determine the cluster for requests on this route.
+    // The cluster provider name must be defined in the associated
+    // :ref:`cluster providers <envoy_v3_api_field_config.route.v3.RouteConfiguration.cluster_providers>`
+    // in the :ref:`name <envoy_v3_api_field_config.core.v3.TypedExtensionConfig.name>` field.
+    string cluster_provider_name = 37;
+
+    // Custom cluster provider configuration to use to determine the cluster for requests on this route.
+    core.v3.TypedExtensionConfig cluster_provider = 39;
```

Comment:
should these two in `oneof`?

Reply:
@lizan 

As you can see, this PR is base on exist unimplemented API. In the original API design, `cluster_provider_name` (original name is `cluster_specifier_plugin`) is used to ref the `cluster_providers` (with type `repeated core.v3.TypedExtensionConfig`) in the `RouteConfiguration`. By this way, mutiple routes can shared same cluster provider to reduce redundant config.

The problem with the exist API design is that when the cluster provider configuration is updated, the entire `RouteConfiguration` must be updated. It is not friendly to users who use `vhds`. And It also increases the complexity of understanding the API. 

So I added an additional inline `cluster_provider` to `Route` to solve these problems. 

From my personal point of view. I think that multiple routes sharing a `cluster_provider` in `RouteConfiguration` can reduce the amount of configuration though. But this advantage should not be obvious. Because the `cluster_provider` itself is a high-level configuration that will rarely be used.

**Ultimately though, the community's opinion is the main one. Should I keep them or just one of them?**

Reply:
I meant you should put those two in a `oneof`. What's the case that you need to specify both `cluster_provider_name` and `cluster_provider` in the same RouteAction?

Reply:
They are already in the `oneof cluster_specifier`. 🤣

Reply:
This is actually the exact opposite of the way we intentionally structured this API in #16944.  We have use-cases where we are going to want multiple routes to use the same cluster specifier plugin, and we do not want to have to repeat the entire plugin config here.

If the problem you're trying to solve here is interoperability with VHDS, then I think a better approach would be to add a `repeated ClusterSpecifierPlugin cluster_specifier_plugins` to the `VirtualHost` message, so that you can update plugins via VHDS.  The requirement would be that any plugin name must be unique across both the `RouteConfiguration` and all `VirtualHost`s within that `RouteConfiguration` (regardless of whether or not they are sent via VHDS).

Reply:
> This is actually the exact opposite of the way we intentionally structured this API in #16944. We have use-cases where we are going to want multiple routes to use the same cluster specifier plugin, and we do not want to have to repeat the entire plugin config here.


So the `oneof` is used. Users can choose config the RouteConfiguration level plugins/providers and ref it in the `Route`, or just config a plugin/provider in the `Route` directly.


The ref way reduces redundant configuration, but makes the API more complex. Most Envoy users use `cluster`. plugin/provider is an advanced configuration that is rarely used, so there won't be too much redundant configuration. I want these users to be able to configure and use the plugin/provider in an easy and intuitive way.

Also, **I hope that eventually plugin/provider will gradually replace `cluster_header` and `weighted_cluster`. At this point, there will be a lot of route-specifed plugin/provider config.** At that time, it was more natural to configure plugin/provider directly in Route.

Reply:
> I also agree with you that most xDS users will use cluster instead of any of the alternatives. However, given that the plugin will be a rarer case, I don't think it's so important for the API to be as simple as possible. I don't think there's really much complexity involved in defining the plugin config out of band even if you don't need it in multiple routes, and I think it makes more sense to have those few users that need plugins deal with that than to add the complexity to the API of having two ways to configure this.

But if no a route-level plugin, the current cluster_header/weighted_cluster configuration will be difficult to migrate.
**Although technically possible, in practice it would be difficult to accept the splitting of route-level cluster_header/weighted_cluster into a route-level plugin name + RouteConfiguration plugin.**

For some control surfaces, such as istio, their logic would be further complicated by such a split.

Reply:
> **Although technically possible, in practice it would be difficult to accept the splitting of route-level cluster_header/weighted_cluster into a route-level plugin name + RouteConfiguration plugin.** For some control surfaces, such as istio, their logic would be further complicated by such a split.

Can you say more about why this would complicate the logic?  It seems like a system like Istio could simply define a convention for the name of a `cluster_header` plugin (e.g., for a header named "foo", the plugin instance name could be "cluster_header_foo"), define that plugin once in the `RouteConfiguration`, and then refer to it in any route that it wants.  I'd like to understand why that's complicated.  The only potential issue I see here is that VHDS could complicate this, because the top-level `RouteConfiguration` will not be in the same resource as the individual routes, but I think the suggestion I offered earlier of adding a list of plugins at the `VirtualHost` level would address that.

Note that there are other reasons why this kind of plugin may need to be defined just once globally.  In the particular use-case that we added this for, the plugin will actually be stateful (it involves communication with an external control plane and a cache), so it would not actually work to have it defined in each individual route, because that would result in each route having a separate instance of the plugin instead of one instance that is used by all routes.  I am concerned that if we support defining a separate plugin instance for each route that the extension API will become more complicated by needing to support two fairly different use-cases.

Reply:
Sorry for the delayed reply. If there is no plugin at the route level, when we migrate from cluster_header/weighted_clusters to the new API, the completely independent Route will inversely depend on RouteConfiguration.

Specifically in istio, the original Virtual Service -> Route is a simple direct mapping relationship. After using the new API, a Virtual Service needs to be converted to Route and some configuration in RouteConfiguration.

Reply:
Sorry, I didn't follow that explanation.  Let me try asking a more specific question: If we allow the plugins to be defined at the `VirtualHost` level so that they work with VHDS, what will not work?

To be clear, I'm really not comfortable with defining the plugin in the individual route.  I think that will badly confuse the semantics of the `ClusterSpecifierPlugin`, which are intended to allow the plugin instance to be stateful and to have that same state apply across multiple routes that are using the same plugin.  I think that in order to consider doing that, we need some justification as to why it's worth introducing that problem, and at the moment, it's not clear to me that we have such a justification.